llvm/lib/Transforms/Vectorize/VectorCombine.cpp

```diff
@@ -729,6 +730,96 @@ bool VectorCombine::foldBitcastShuf(Instruction &I) {
   return true;
 }
 
+/// VP Intrinsics whose vector operands are both splat values may be simplified
+/// into the scalar version of the operation and the result is splatted. This
+/// can lead to scalarization down the line.
+bool VectorCombine::scalarizeVPIntrinsic(VPIntrinsic &VPI) {
+  Value *Op0 = VPI.getArgOperand(0);
+  Value *Op1 = VPI.getArgOperand(1);
+
+  if (!getSplatValue(Op0) || !getSplatValue(Op1))
+    return false;
+
+  // For the binary VP intrinsics supported here, the result on disabled lanes
+  // is a poison value. For now, only do this simplification if all lanes
+  // are active.
+  // TODO: Relax the condition that all lanes are active by using insertelement
+  // on inactive lanes.
+  auto IsAllTrueMask = [](Value *MaskVal) {
+    if (Value *SplattedVal = getSplatValue(MaskVal))
+      if (auto *ConstValue = dyn_cast<Constant>(SplattedVal))
+        return ConstValue->isAllOnesValue();
+    return false;
+  };
+  if (!IsAllTrueMask(VPI.getArgOperand(2)))
+    return false;
```

Comment:
I tried with this simplification and it is unable to determine that the allonemask is in fact all ones. It is because `isAllOnesValue` does not do the dyn_cast to constant. I added the dyn_cast to `isAllOnesValue` and it correctly identifies it as a splat.

Reply:
This is how it is done in `ExpandVectorPredication::isAllTrueMask`

Reply:
Ah ok, fair enough.